File: 55.jump-game.py
#
# @lc app=leetcode id=55 lang=python3
#
# [55] Jump Game
#

# @lc code=start
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reach = 0
        i = 0
        
        while i < len(nums):
            if i > reach:
                return False
            reach = max(reach , i + nums[i] )
            i = i + 1
        return True  

# @lc code=end

File: golang/55.jump-game.go
/*
 * @lc app=leetcode.cn id=55 lang=golang
 *
 * [55] Jump Game
 */
package main

import "fmt"

// @lc code=start
func canJump(nums []int) bool {
	dp := make([]bool, len(nums))
	dp[0] = true
	for i := 0; i < len(nums); i++ {
		if !dp[i] {
			return false
		}
		for j := i; j <= i+nums[i]; j++ {
			if j < len(nums) {
				dp[j] = true
			}
		}
	}
	return dp[len(dp)-1]
}

// @lc code=end
func main() {
	fmt.Println(canJump([]int{2, 3, 1, 1, 4}))
}
